cohortextractor/main.py

```diff
@@ -32,18 +37,102 @@ def generate_cohort(
         definition_path=definition_path,
         output_file=output_file,
         backend=backend_id,
+        index_date_range=index_date_range,
         dummy_data_file=dummy_data_file,
     )
 
-    cohort = load_cohort(definition_path)
     output_file.parent.mkdir(parents=True, exist_ok=True)
-    if dummy_data_file and not db_url:
-        validate_dummy_data(cohort, dummy_data_file, output_file)
-        shutil.copyfile(dummy_data_file, output_file)
+
+    for index_date in _generate_date_range(index_date_range):
+        if index_date is not None:
+            log.info(f"Setting index_date to {index_date}")
+            date_suffix = f"_{index_date}"
+        else:
+            date_suffix = ""
+        cohort = load_cohort(definition_path, index_date)
+        output_file_with_date = Path(str(output_file).replace("*", date_suffix))
+        if dummy_data_file and not db_url:
+            dummy_data_file_with_date = Path(
+                str(dummy_data_file).replace("*", date_suffix)
+            )
+            validate_dummy_data(
+                cohort, dummy_data_file_with_date, output_file_with_date
+            )
+            shutil.copyfile(dummy_data_file_with_date, output_file_with_date)
+        else:
+            backend = BACKENDS[backend_id](
+                db_url, temporary_database=temporary_database
+            )
+            results = extract(cohort, backend)
+            write_output(results, output_file_with_date)
+
+
+def _generate_date_range(date_range_str):
+    # Bail out with an "empty" range: this means we don't need separate
+    # codepaths to handle the range, single date, and no date supplied cases
+    if not date_range_str:
+        return [None]
+    start, end, period = _parse_date_range(date_range_str)
+    if end < start:
+        raise ValueError(
+            f"Invalid date range '{date_range_str}': end cannot be earlier than start"
+        )
+    dates = []
+    while start <= end:
+        dates.append(start.isoformat())
+        start = _increment_date(start, period)
+    # The latest data is generally more interesting/useful so we may as well
+    # extract that first
+    dates.reverse()
+    return dates
+
+
+def _parse_date_range(date_range_str):
+    period = "month"
+    if " to " in date_range_str:
+        start, end = date_range_str.split(" to ", 1)
+        if " by " in end:
+            end, period = end.split(" by ", 1)
+    else:
+        start = end = date_range_str
+    try:
+        start = _parse_date(start)
+        end = _parse_date(end)
+    except ValueError:
+        raise ValueError(
+            f"Invalid date range '{date_range_str}': Dates must be in YYYY-MM-DD "
+            f"format or 'today' and ranges must be in the form "
+            f"'DATE to DATE by (week|month)'"
```

Comment:
I think if I got this error I would interpret it to mean that the `by` clause was mandatory.

Reply:
This is how it currently is in V1, but I agree.  The error message could be more specific to what's actually idenitified as wrong